File: host-console/src/lib/tec-calibration.ts
export interface TecCalibrationPoint {
  tempC: number
  tecVoltageV: number
  wavelengthNm: number
}

const measuredTempVoltagePoints = [
  { tempC: 5.0, tecVoltageV: 0.11 },
  { tempC: 6.0, tecVoltageV: 0.18 },
  { tempC: 7.5, tecVoltageV: 0.221 },
  { tempC: 10.7, tecVoltageV: 0.316 },
  { tempC: 15.5, tecVoltageV: 0.473 },
  { tempC: 20.0, tecVoltageV: 0.606 },
  { tempC: 20.5, tecVoltageV: 0.63 },
  { tempC: 28.2, tecVoltageV: 0.957 },
  { tempC: 29.0, tecVoltageV: 0.985 },
  { tempC: 37.4, tecVoltageV: 1.345 },
  { tempC: 44.7, tecVoltageV: 1.64 },
  { tempC: 46.1, tecVoltageV: 1.7 },
  { tempC: 54.5, tecVoltageV: 2.03 },
  { tempC: 58.7, tecVoltageV: 2.182 },
  { tempC: 65.0, tecVoltageV: 2.429 },
]

function clamp(value: number, min: number, max: number): number {
  if (value < min) {
    return min
  }

  if (value > max) {
    return max
  }

  return value
}

function interpolate(points: readonly { x: number; y: number }[], value: number): number {
  if (points.length === 0) {
    return 0
  }

  if (value <= points[0].x) {
    return points[0].y
  }

  const lastPoint = points[points.length - 1]
  if (value >= lastPoint.x) {
    return lastPoint.y
  }

  for (let index = 1; index < points.length; index += 1) {
    const previous = points[index - 1]
    const current = points[index]

    if (value <= current.x) {
      const span = current.x - previous.x
      const ratio = span === 0 ? 0 : (value - previous.x) / span
      return previous.y + (current.y - previous.y) * ratio
    }
  }

  return lastPoint.y
}

function inferTempFromVoltage(tecVoltageV: number): number {
  const table = measuredTempVoltagePoints
    .map((point) => ({ x: point.tecVoltageV, y: point.tempC }))
    .sort((left, right) => left.x - right.x)

  return interpolate(table, tecVoltageV)
}

// The screenshot provided by the user contains wavelength rows where the
// temperature label is omitted. Those rows are recovered here by inferring
// temperature from the measured TEC voltage using the same calibration table.
export const tecCalibrationPoints: TecCalibrationPoint[] = [
  { tempC: 5.0, tecVoltageV: 0.11, wavelengthNm: 771.2 },
  { tempC: 7.5, tecVoltageV: 0.221, wavelengthNm: 771.8 },
  { tempC: 10.7, tecVoltageV: 0.316, wavelengthNm: 772.8 },
  { tempC: 15.5, tecVoltageV: 0.473, wavelengthNm: 774.8 },
  { tempC: 20.0, tecVoltageV: 0.606, wavelengthNm: 776.8 },
  { tempC: 20.5, tecVoltageV: 0.63, wavelengthNm: 776.9 },
  { tempC: inferTempFromVoltage(0.8), tecVoltageV: 0.8, wavelengthNm: 778.4 },
  { tempC: 28.2, tecVoltageV: 0.957, wavelengthNm: 779.8 },
  { tempC: 29.0, tecVoltageV: 0.985, wavelengthNm: 780.1 },
  { tempC: inferTempFromVoltage(1.224), tecVoltageV: 1.224, wavelengthNm: 780.5 },
  { tempC: 37.4, tecVoltageV: 1.345, wavelengthNm: 781.6 },
  { tempC: inferTempFromVoltage(1.511), tecVoltageV: 1.511, wavelengthNm: 781.6 },
  { tempC: 44.7, tecVoltageV: 1.64, wavelengthNm: 783.0 },
  { tempC: 46.1, tecVoltageV: 1.7, wavelengthNm: 783.1 },
  { tempC: inferTempFromVoltage(1.8), tecVoltageV: 1.8, wavelengthNm: 783.2 },
  { tempC: inferTempFromVoltage(1.9), tecVoltageV: 1.9, wavelengthNm: 784.2 },
  { tempC: 54.5, tecVoltageV: 2.03, wavelengthNm: 784.6 },
  { tempC: 58.7, tecVoltageV: 2.182, wavelengthNm: 786.1 },
  { tempC: inferTempFromVoltage(2.235), tecVoltageV: 2.235, wavelengthNm: 787.4 },
  { tempC: 65.0, tecVoltageV: 2.429, wavelengthNm: 790.0 },
].sort((left, right) => left.tempC - right.tempC)

const wavelengthByTempTable = tecCalibrationPoints.map((point) => ({
  x: point.tempC,
  y: point.wavelengthNm,
}))

const tempByWavelengthTable = tecCalibrationPoints
  .map((point) => ({ x: point.wavelengthNm, y: point.tempC }))
  .sort((left, right) => left.x - right.x)

const voltageByTempTable = tecCalibrationPoints.map((point) => ({
  x: point.tempC,
  y: point.tecVoltageV,
}))

export function clampTecTempC(tempC: number): number {
  return clamp(tempC, tecCalibrationPoints[0].tempC, tecCalibrationPoints[tecCalibrationPoints.length - 1].tempC)
}

export function clampTecVoltageV(tecVoltageV: number): number {
  return clamp(
    tecVoltageV,
    tecCalibrationPoints[0].tecVoltageV,
    tecCalibrationPoints[tecCalibrationPoints.length - 1].tecVoltageV,
  )
}

export function clampTecWavelengthNm(wavelengthNm: number): number {
  const min = tempByWavelengthTable[0].x
  const max = tempByWavelengthTable[tempByWavelengthTable.length - 1].x
  return clamp(wavelengthNm, min, max)
}

export function estimateWavelengthFromTempC(tempC: number): number {
  return interpolate(wavelengthByTempTable, clampTecTempC(tempC))
}

export function estimateTempFromWavelengthNm(wavelengthNm: number): number {
  return interpolate(tempByWavelengthTable, clampTecWavelengthNm(wavelengthNm))
}

export function estimateTempFromTecVoltageV(tecVoltageV: number): number {
  return inferTempFromVoltage(clampTecVoltageV(tecVoltageV))
}

export function estimateTecVoltageFromTempC(tempC: number): number {
  return interpolate(voltageByTempTable, clampTecTempC(tempC))
}

export function estimateTecVoltageFromWavelengthNm(wavelengthNm: number): number {
  return estimateTecVoltageFromTempC(estimateTempFromWavelengthNm(wavelengthNm))
}

export function estimateWavelengthFromTecVoltageV(tecVoltageV: number): number {
  return estimateWavelengthFromTempC(estimateTempFromTecVoltageV(tecVoltageV))
}

File: ios/BSLProtocol/Sources/BSLProtocol/Models/TecCalibration.swift
import Foundation

/// Wavelength <-> temperature <-> TEC-voltage LUT.
///
/// Swift port of `host-console/src/lib/tec-calibration.ts`. Kept in
/// `BSLProtocol` so the host console and the iOS app share one authoritative
/// calibration curve — see `docs/protocol-spec.md` and the firmware default
/// LUT in `components/laser_controller/src/laser_controller_config.c:7-43`.
///
/// The data points were measured on the bench. Rows where temperature was
/// not directly captured are recovered by inferring from the TEC voltage
/// against the voltage-temperature sub-table.
public enum TecCalibration {

    public struct Point: Sendable, Equatable {
        public let tempC: Double
        public let tecVoltageV: Double
        public let wavelengthNm: Double
    }

    private static let tempVoltagePoints: [(tempC: Double, tecVoltageV: Double)] = [
        (5.0, 0.11),
        (6.0, 0.18),
        (7.5, 0.221),
        (10.7, 0.316),
        (15.5, 0.473),
        (20.0, 0.606),
        (20.5, 0.63),
        (28.2, 0.957),
        (29.0, 0.985),
        (37.4, 1.345),
        (44.7, 1.64),
        (46.1, 1.7),
        (54.5, 2.03),
        (58.7, 2.182),
        (65.0, 2.429),
    ]

    /// Calibration points sorted by temperature. Source of truth for every
    /// conversion below.
    public static let points: [Point] = {
        let rawPoints: [(tempC: Double?, tecVoltageV: Double, wavelengthNm: Double)] = [
            (5.0, 0.11, 771.2),
            (7.5, 0.221, 771.8),
            (10.7, 0.316, 772.8),
            (15.5, 0.473, 774.8),
            (20.0, 0.606, 776.8),
            (20.5, 0.63, 776.9),
            (nil, 0.8, 778.4),
            (28.2, 0.957, 779.8),
            (29.0, 0.985, 780.1),
            (nil, 1.224, 780.5),
            (37.4, 1.345, 781.6),
            (nil, 1.511, 781.6),
            (44.7, 1.64, 783.0),
            (46.1, 1.7, 783.1),
            (nil, 1.8, 783.2),
            (nil, 1.9, 784.2),
            (54.5, 2.03, 784.6),
            (58.7, 2.182, 786.1),
            (nil, 2.235, 787.4),
            (65.0, 2.429, 790.0),
        ]
        return rawPoints
            .map { raw in
                Point(
                    tempC: raw.tempC ?? inferTempFromVoltage(raw.tecVoltageV),
                    tecVoltageV: raw.tecVoltageV,
                    wavelengthNm: raw.wavelengthNm
                )
            }
            .sorted { $0.tempC < $1.tempC }
    }()

    /// Clamp `tempC` to the calibration window.
    public static func clampTempC(_ tempC: Double) -> Double {
        clamp(tempC, points.first?.tempC ?? 0, points.last?.tempC ?? 0)
    }

    /// Clamp `wavelengthNm` to the calibration window.
    public static func clampWavelengthNm(_ wavelengthNm: Double) -> Double {
        let sorted = points.sorted { $0.wavelengthNm < $1.wavelengthNm }
        return clamp(wavelengthNm, sorted.first?.wavelengthNm ?? 0, sorted.last?.wavelengthNm ?? 0)
    }

    /// Estimate the TEC target temperature (°C) for a given wavelength (nm).
    /// Uses linear interpolation across the bench-measured LUT. Mirrors the
    /// firmware wavelength LUT exactly.
    public static func tempFromWavelengthNm(_ wavelengthNm: Double) -> Double {
        let table = points
            .map { ($0.wavelengthNm, $0.tempC) }
            .sorted { $0.0 < $1.0 }
        return interpolate(table, clampWavelengthNm(wavelengthNm))
    }

    /// Estimate the wavelength (nm) that corresponds to a given target
    /// temperature (°C).
    public static func wavelengthFromTempC(_ tempC: Double) -> Double {
        let table = points
            .map { ($0.tempC, $0.wavelengthNm) }
            .sorted { $0.0 < $1.0 }
        return interpolate(table, clampTempC(tempC))
    }

    /// Estimate the TEC command voltage for a given target temperature (°C).
    public static func tecVoltageFromTempC(_ tempC: Double) -> Double {
        let table = points
            .map { ($0.tempC, $0.tecVoltageV) }
            .sorted { $0.0 < $1.0 }
        return interpolate(table, clampTempC(tempC))
    }

    private static func inferTempFromVoltage(_ voltage: Double) -> Double {
        let table = tempVoltagePoints
            .map { ($0.tecVoltageV, $0.tempC) }
            .sorted { $0.0 < $1.0 }
        return interpolate(table, voltage)
    }

    private static func interpolate(_ table: [(Double, Double)], _ value: Double) -> Double {
        guard !table.isEmpty else { return 0 }
        if value <= table[0].0 { return table[0].1 }
        if let last = table.last, value >= last.0 { return last.1 }
        for i in 1..<table.count {
            let prev = table[i - 1]
            let cur = table[i]
            if value <= cur.0 {
                let span = cur.0 - prev.0
                let ratio = span == 0 ? 0 : (value - prev.0) / span
                return prev.1 + (cur.1 - prev.1) * ratio
            }
        }
        return table.last?.1 ?? 0
    }

    private static func clamp(_ v: Double, _ lo: Double, _ hi: Double) -> Double {
        max(lo, min(hi, v))
    }
}
